defi/single-token-liquidity-pool-2/src/lib.rs

```diff
@@ -0,0 +1,132 @@
+use scrypto::prelude::*;
+
+blueprint! {
+    struct LiquidityPool {
+        pool: Vault,
+
+        lp_mint_badge: Vault,
+        lp_resource_address: ResourceAddress,
+        total_lp: Decimal,
+        total_liquidity: Decimal,
+        lp_divisibility: u8,
+    }
+
+    impl LiquidityPool {
+        /// Creates a LiquidityPool component and returns the component address
+        /// along with the initial LP tokens.
+        pub fn new(
+            tokens: Bucket,
+            lp_symbol: String,
+            lp_name: String,
+            lp_divisibility: u8,
+        ) -> (ComponentAddress,Bucket) {
+
+            // Check arguments
+            assert!(
+                !tokens.is_empty(),
+                "You must pass in an initial supply of token"
+            );
+
+            // Mint the badge needed to mint LP Tokens
+            let lp_mint_badge = ResourceBuilder::new_fungible()
+                .divisibility(DIVISIBILITY_NONE)
+                .metadata("name", "LP Token Mint Auth")
+                .initial_supply(1);
+
+            //  Define our LP token and mint an initial supply
+            let lp_tokens = ResourceBuilder::new_fungible()
+                .divisibility(lp_divisibility)
+                .metadata("symbol", lp_symbol)
+                .metadata("name", lp_name)
+                .mintable(rule!(require(lp_mint_badge.resource_address())), LOCKED)
+                .burnable(rule!(require(lp_mint_badge.resource_address())), LOCKED)
+                .initial_supply(tokens.amount());
+            let lp_resource_address = lp_tokens.resource_address();
+
+            let total_lp  = tokens.amount();
+            let total_liquidity = tokens.amount();
+            
+            let liquidity_pool = Self {
+                pool: Vault::with_bucket(tokens),
+
+                lp_mint_badge: Vault::with_bucket(lp_mint_badge),
+                lp_resource_address,
+
+                total_lp,
+                total_liquidity,
+                lp_divisibility,
+            }
+            .instantiate()
+            .globalize();
+
+            // Return the new LiquidityPool component, as well as the initial supply of LP tokens
+            (liquidity_pool,lp_tokens)
+        }
+
+        /// Adds liquidity to this pool and return the LP tokens representing pool shares
+        /// along with any remainder.
+        pub fn add_liquidity(&mut self,
+             mut tokens: Bucket
+        ) -> (Bucket,Bucket) {
```

Comment:
The other bucket returns should no longer be required one you make the change:

`self.pool.put(tokens.take(tokens.amount()))` => `self.pool.put(tokens)`

```suggestion
        ) -> Bucket {
```